File: ml-benchmark/synduce/constraints/ensures/mtss_full_noe.f
Inductive List = elt Int | cons {Int, List};
Inductive NList = line List | ncons {List, NList};
Inductive CNList = sglt List | cat {CNList, CNList};

cton = fix (
  \f: CNList -> NList.
  let dec = fix (
    \g: CNList -> CNList -> NList. \l: CNList. \c: CNList.
    match c with
      sglt x -> ncons {x, f l}
    | cat {x, y} -> g (cat {y, l}) x
    end
  ) in \c: CNList.
  match c with
    sglt x -> line x
  | cat {x, y} -> dec y x
  end
);

sum = fix (
  \f: List -> Int. \xs: List.
  match xs with
    elt x -> x
  | cons {h, t} -> + h (f t)
  end
);

max = \a: Int. \b: Int. if < a b then b else a;

spec = \xs: NList. (fix (
  \f: NList -> {Int, Int}. \xs: NList.
  match xs with
    line a -> {max 0 (sum a), sum a}
  | ncons {h, t} ->
    let hsum = sum h in
      let tres = f t in
        {max (+ tres.1 hsum) 0, + tres.2 hsum}
  end
) xs).1;

target = fix (
  \f: CNList -> Compress CNList.
  let tsum = fix (
    \g: List -> Compress List. \xs: List.
    match xs with
      elt x -> elt x
    | cons {h, t} -> cons {h, g t}
    end
  ) in \c: CNList.
  match c with
    sglt x -> sglt (tsum x)
  | cat {l, r} -> cat {f l, f r}
  end
);

main = \c: CNList. spec (cton (target c));

File: ml-benchmark/synduce/constraints/ensures/mtss_full_noe.ml
type list_ = Elt of int | Cons of int * list_
type nlist = Line of list_ | Ncons of list_ * nlist
type cnlist = Sglt of list_ | Cat of cnlist * cnlist

let rec cton c =
  let rec dec l c2 =
    match c2 with
    | Sglt x -> Ncons (x, cton l)
    | Cat (x, y) -> dec (Cat (y, l)) x
  in
  match c with
  | Sglt x -> Line x
  | Cat (x, y) -> dec y x

let rec sum xs =
  match xs with
  | Elt x -> x
  | Cons (h, t) -> h + sum t

let max a b = if a < b then b else a

let spec xs =
  let rec f ys =
    match ys with
    | Line a -> (max 0 (sum a), sum a)
    | Ncons (h, t) ->
        let hsum = sum h in
        match f t with
        | (u1, u2) -> (max (u1 + hsum) 0, u2 + hsum)
  in
  match f xs with
  | (u, _) -> u

let rec target c =
  let rec tsum xs =
    match xs with
    | Elt x -> Elt x
    | Cons (h, t) -> Cons (h, tsum t)
  in
  match c with
  | Sglt x -> Sglt (tsum x)
  | Cat (l, r) -> Cat (target l, target r)

let program c = spec (cton (target c))